Create How-to-Build-a-Flask-ReactJS-Application-and-Deploy-to-Okteto.md (#40)
okteto/blog-proposals: How-to-Build-a-Flask-ReactJS-Application-and-Deploy-to-Okteto.md

# How to Build a Flask, ReactJS Application and Deploy to Okteto
## Introduction 

**[Flask](https://flask.palletsprojects.com/en/2.0.x/)** is a great open source web microframework built with Python. It is straightforward and easy to understand.
Flask is used to build web applications and also APIs. In this article we will use it to build an API that displays date and time.

> ReactJS is an open-source  front-end  JavaScript library used to build user interfaces.

Reviewer: Can you add a link to the ReactJS website?

> **[Okteto](https://okteto.com/)** is an open-source tool that runs locally to synchronize application code changes to a running pod in a local or remote Kubernetes cluster. With Okteto we are able to see deployment version of our app as we are developing it.

Reviewer: ```suggestion
**[Okteto](https://okteto.com/)** is an open-source tool that runs locally to synchronize application code changes to a running pod in your Kubernetes cluster. With Okteto we are able to see the deployment version of our app during the development process.
```

Okteto has a development environment which we are going to use instead of working locally. Of course we can build this project with our development but we’ll be using Okteto's development environment to build our app which gives us a production-like development environment.

> In this article we will build an application that displays the date and time with React and Flask and deployment will be handled with Okteto

Reviewer: Do you mind building another sample for this draft? A to-do list application or an application with a bit of complexity will be better off than a date and time display application.

Author: I stated the project I will be building in the outline but there were no queries.
But no worries, I'll make the appropriate changes.

Reviewer: Hi, I'm sorry I might have missed the changes you made in the outline. Thank you.

On your text editor, navigate to `frontend` folder and create `Dockerfile` and input the code:

```Dockerfile
 
FROM node:14.17.3

WORKDIR /user/src/app 

COPY ./package.json ./

COPY ./package-lock.json ./

RUN npm install 

COPY . .

EXPOSE 8080

ENTRYPOINT [ "npm" ]

CMD ["start"]

```

I know you would like to test your code, but let's first create your `docker-compose.yml` file so that you can run both servers at once.
With the `docker-compose.yml` commands you will connect the images created for React and Flask and run them together. You will also use the `docker-compose.yml` to set how you would like Okteto to run your application.
In `flask-react-app` folder create `docker-compose.yml` file and put the code:

```yml
services:
  api:
    build:  ./api
    ports: 
    - "5000:5000"
    volumes: 
    - ./api:/api
    environment: 
      FLASK_ENV: development
  web:
    build: ./frontend
    ports: 
      - "8080:8080"
    volumes: 
      - ./frontend:/user/src/app
    depends_on: 
      - api

```
Now run `$ docker-compose build` and then `$ docker-compose up`

You'll see that your react application is running on port `localhost:8080`

> ![](First-image)

Reviewer: Can you provide an image of your frontend application running at localhost:8080?

Author: it's the default page that comes when you run `$ npx create-react-app`, which will be impossible to get unless I go through the steps from the beginning.

import "./App.css";
import { TodoPage } from "./components/TodoList";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { Detail } from "./components/detailView";
// import { Form } from './components/form';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <Router>
          <Switch>
            <Route exact path="/">
              <TodoPage />
            </Route>
            <Route path="/:id">
              <Detail />
            </Route>
          </Switch>
        </Router>
      </header>
    </div>
  );
}

export default App;
```
There are some tools I used but haven't installed yet, so what you will do is update your `package.json` file to look like below:
```json
{
  "name": "frontend",
  "version": "0.1.0",
  "proxy": "http://api:5000",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^11.2.7",
    "@testing-library/user-event": "^12.8.3",
    "react": "^16.13.1",
    "react-dom": "^17.0.2",
    "react-router": "5.2.0",
    "react-router-dom": "^5.2.0",
    "react-scripts": "4.0.3",
    "web-vitals": "^1.1.2"
  },
  ...

> ```

Reviewer: Can you list the packages you used? It's unethical to manually add packages.

Please send a command such as `npm install x y z`

Reviewer: @khabdrick, can you fix this?

Author: My bad, I thought I already committed the change
I have fixed it now!
